### PR TITLE
Improve handling of dllexport

### LINKED ISSUE #7196: Add support for DllExport on Windows
It currently does not seem to be possible to export a function from a DLL on Windows. It's not done as one would expect when building with --lib.

### PR DESCRIPTION
**Pull request by alexcrichton**
These two commits are aimed at "fixing" our usage of `dllexport` in the compiler. Currently we blanket apply this attribute to _everything_ public in a crate, but this ends up with a few downsides:
- Executables are larger than the should be as a result of thinking they should export everything
- Native libraries aren't handled correctly because technically a statically included native library should be exported from a DLL in some cases.
- Symbols don't actually need to be exported if they never end up in a DLL.

The first commit adds a new unstable attribute, `#[linked_from]`, which is a way to tell the compiler what native library a block of symbols comes from. This is used to inform the compiler what set of native libraries are statically included in the rlib (or other output). This information is later used to export them from a DLL if necessary. Currently this is only used in a few places (such as the LLVM bindings) to get the compiler to link correctly.

The second commit stops adding `dllexport` to all items in LLVM and instead explicitly telling the linker what symbols should be exported. We only need to do this when building a dynamic library, and otherwise we can avoid adding `dllexport` or telling the linker about exported symbols.

As a testament to this change, the size of "Hello World" on MSVC drops from 1.2MB to 67KB as a result of this patch. This is because the linker can much more aggressively remove unused code.

These commits do not yet attempt to fix our story with `dllimport`, and I'll leave that to a future PR and issue, for now though I'm going to say that this

Closes #7196

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @brson 

cc @retep998

**Comment by retep998** [comment]
>As a testament to this change, the size of "Hello World" on MSVC drops from 1.2MB to 67KB as a result of this patch.

:+1:

**Comment by alexcrichton** [comment]
Rebased, ping r? @brson 

---

@retep998 

>  EDIT: Tried building this myself. Doesn't seem to be any discernible difference for Rust's DLLs, but regular binaries produced by Rust with statically linked libraries benefit massively.

Yeah right now we don't have a great distinction between "dylib later linked into a Rust project" and "dylib only used to link into an external C-like project", so we have to continue to pessimistically export everything. If we add this distinction, however, then we could get the same wins on dylibs as we're getting on executables.

I'd also recommend avoiding editing github comments, unfortunately no notifications are sent out after an edit...

**Comment by brson** [comment]
This will break dylibs on windows that, like rustllvm, export native symbols, right?

**Comment by alexcrichton** [comment]
Thankfully this doesn't actually break any of them that weren't already broken. If the native library is compiled with `dllexport` then it'll still continue to be exported, and if it wasn't compiled with `dllexport` it will continue to not be exported.

LLVM doesn't support a mode of being built with `dllexport`, so we used a custom `*.def` file to export them previously, but the linker only accepts one `*.def` file and now that the compiler is providing its own we needed a mechanism for the compiler to emit those symbols as well.

In short, this doesn't actually change the story for any native library linked into an MSVC binary today. It still sucks that you have to remember to `dllexport` it yourself if you know it's going to be called through a DLL, but that's true both before and after this PR.

**Comment by retep998** [comment]
@alexcrichton If we really want some wins with dynamic libraries we should respect the distinction between `.dll` and `.lib` where when Rust produces a dylib on Windows it'll produce both a `.lib` with all the Rust metadata and stuff for linking to and a `.dll` which only contains fully compiled and linked code rather than the current situation of the `.dll` containing both.

**Comment by alexcrichton** [comment]
@retep998 that would only help binary size, right? I'm not too too worried about it because Rust dlls are so uncommon. This is another case where a distinction between a DLL-used-with-a-C-api and a DLL-used-as-a-rust-dependency would be useful, however.

**Comment by retep998** [comment]
@alexcrichton It would help binary size when you are distributing DLLs with a Rust binary where nobody will link to those DLLs. For example the DLLs sitting next to `rustc.exe` are never linked to by users of the Rust distribution, so there is no need for them to have all that metadata.

**Comment by alexcrichton** [comment]
Sure, but this is nothing that has to do specifically with Windows, this is a problem that plagues all platforms.
